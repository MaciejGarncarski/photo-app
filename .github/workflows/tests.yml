name: Tests
on:
  push:
    branches:
      - main
      - develop
  pull_request: {}

jobs:
  test:
    name: 🃏 Test
    runs-on: ubuntu-latest
    steps:
      - name: 🛑 Cancel Previous Runs
        uses: styfle/cancel-workflow-action@0.11.0

      - name: ⬇️ Checkout repo
        uses: actions/checkout@v3
        with:
          persist-credentials: false      

      - name: ⎔ Set up node
        uses: actions/setup-node@v3
        with:
          node-version: lts/*

      - uses: pnpm/action-setup@v2
        name: Install pnpm
        with:
          version: 8
          run_install: false

      - name: Get pnpm store directory
        shell: bash
        run: |
          echo "STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV

      - uses: actions/cache@v3
        name: Setup pnpm cache
        with:
          path: ${{ env.STORE_PATH }}
          key: ${{ runner.os }}-pnpm-store-${{ hashFiles('**/pnpm-lock.yaml') }}
          restore-keys: |
            ${{ runner.os }}-pnpm-store-

      - name: 📥 Download deps
        run: pnpm install
        
      - name: 🔎 Type check
        run: pnpm tsc

      - name: 🧪 Test
        run: pnpm coverage