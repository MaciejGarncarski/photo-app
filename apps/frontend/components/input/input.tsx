import clsx from 'clsx'
import { type ChangeEvent, Ref } from 'react'

import styles from './input.module.scss'

type Props = {
	labelText: string
	placeholder: string
	type?: 'text' | 'number' | 'tel' | 'email' | 'password'
	variant: 'primary' | 'secondary'
	error?: string
	optional?: boolean
	value?: string
	ref?: Ref<HTMLInputElement>
	onChange?: (changeEv: ChangeEvent<HTMLInputElement>) => void
}

export const Input = ({
	type = 'text',
	labelText,
	error,
	optional,
	placeholder,
	variant,
	ref,
	...props
}: Props) => {
	const containerClassName = clsx(
		error && styles.containerError,
		styles.container,
	)

	const inputClassName = clsx(
		{
			[styles.inputError]: Boolean(error),
		},
		styles[variant],
		styles.input,
	)

	return (
		<div>
			<div className={containerClassName}>
				<label className={styles.label} htmlFor={labelText}>
					{labelText} {optional && '(optional)'}
				</label>
				<input
					ref={ref}
					type={type}
					id={labelText}
					data-testid="input"
					placeholder={placeholder}
					className={inputClassName}
					{...props}
				/>
			</div>
			{error && <p className={styles.error}>{error}</p>}
		</div>
	)
}
