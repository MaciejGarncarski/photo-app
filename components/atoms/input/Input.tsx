import clsx from 'clsx';
import { ChangeEvent, forwardRef } from 'react';
import { FieldError } from 'react-hook-form';

import styles from './input.module.scss';

type PropsTypes = {
  type?: 'text' | 'number' | 'tel' | 'email' | 'password';
  labelText: string;
  error?: FieldError;
  optional?: boolean;
  className?: string;
  onChange?: (ev: ChangeEvent<HTMLInputElement>) => void;
  value?: string;
};

export const Input = forwardRef<HTMLInputElement, PropsTypes>(
  ({ type = 'text', labelText, error, optional, className, onChange, value, ...props }, ref) => {
    const containerClassName = clsx(className, error && styles.containerError, styles.container);

    const inputClassName = clsx(
      {
        [styles.inputError]: Boolean(error),
      },
      styles.input,
    );

    return (
      <div className={containerClassName}>
        <label className={styles.label} htmlFor={labelText}>
          {labelText} {optional && '(optional)'}
        </label>
        <input
          ref={ref}
          type={type}
          id={labelText}
          data-testid="input"
          className={inputClassName}
          onChange={onChange}
          value={value}
          {...props}
        />
        {error && <p className={styles.error}>{error.message}</p>}
      </div>
    );
  },
);
