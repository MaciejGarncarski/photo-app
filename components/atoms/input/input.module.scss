@use '/styles/common.scss';

.container {
  display: flex;
  flex-direction: column;
  position: relative;
  gap: 0.5rem;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  color: var(--text-secondary);
  text-transform: capitalize;
}


.input {
  background-color: transparent;
  border: none;
  padding: 1rem;
  border: var(--border-gray);
  border-radius: var(--radius-medium);
  transition: all var(--transition);
  transition-property: background-color, box-shadow;

  &:hover {
    background-color: var(--bg-secondary);
  }

  &:focus {
    outline: none;
    border-color: transparent;
    box-shadow: var(--shadow-focus);
  }
}

.inputError {
  background-color: var(--bg-error);
  border-color: var(--color-error);
}


.error {
  color: var(--color-error);
  font-weight: 600;
}

.textarea {
  border: none;
  padding: 1.5rem;
  max-width: calc(100% - 1px);
  resize: vertical;
}