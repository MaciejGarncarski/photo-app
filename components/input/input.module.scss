@use 'common';

.container {
  display: flex;
  flex-direction: column;
  position: relative;
  gap: 0.5rem;
  --gap: 0.7rem;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  color: var(--color-text-gray);
  transition: all var(--transition);
  transition-property: background-color, transform, opacity;
  display: inline-block;
}

.input {
  background-color: var(--color-secondary);
  padding: 1rem var(--gap);
  border-radius: var(--radius-small);
  border: 1px solid var(--shadow-0-5);
  box-shadow: var(--shadow);
  color: var(--color-text);

  &:focus {
    outline: none;
  }
}

.primary {
  background-color: var(--color-secondary);
}

.secondary {
  @include common.mediaquery('desktop') {
    background-color: var(--color-primary);
  }
}

.inputError {
  background-color: var(--color-error-light);

  &::placeholder {
    color: var(--color-text);
    font-weight: 500;
  }
}

.error {
  color: var(--color-error);
  padding-left: var(--gap);
  font-weight: 400;
  margin-left: 0.75rem;
  margin-top: 0.25rem;
}
