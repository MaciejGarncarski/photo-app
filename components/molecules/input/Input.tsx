import clsx from 'clsx';
import { forwardRef } from 'react';
import { FieldError } from 'react-hook-form';

import styles from './input.module.scss';

type InputProps = {
  type?: JSX.IntrinsicElements['input']['type'];
  labelText: string;
  isDirty?: boolean;
  error?: FieldError;
  optional?: boolean;
  className?: string;
};

export const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ type = 'text', labelText, isDirty, error, optional, className, ...props }, ref) => {
    return (
      <div
        className={clsx(className && className, error && styles.containerError, styles.container)}
      >
        <input className={styles.input} ref={ref} type={type} {...props} id={labelText} />
        <label className={clsx(isDirty && styles.labelActive, styles.label)} htmlFor={labelText}>
          {labelText} {optional && '(optional)'}
        </label>
      </div>
    );
  }
);
