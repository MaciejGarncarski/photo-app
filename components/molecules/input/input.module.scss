@use '/styles/common.scss';

.container {
  display: flex;
  flex-direction: column;
  border: 2px solid var(--color-primary);
  border-radius: var(--radius-small);
  position: relative;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  position: absolute;
  left: 0rem;
  top: 0;
  padding: 0 1rem;
  color: var(--text-placeholder);
  transition: all 100ms ease-out;
  cursor: text;
  transition-property: background-color, transform, font-size;
}

.labelActive,
.input:focus+.label {

  background-color: var(--bg-primary);
  color: var(--text-primary);
  top: -1.25rem;
  left: 0.5rem;


  @include common.mediaquery('lg') {
    font-size: 2.5rem;
  }

  @include common.mediaquery('xl') {
    top: -2rem;
    left: 1rem;
  }
}

.input {
  background-color: transparent;
  border: none;
  padding: 1rem;
}

.error {
  color: var(--color-error);
  font-weight: 600;
  font-size: 1.5rem;
  margin-left: 1rem;

  @include common.mediaquery('lg') {
    font-size: 2rem;
  }
}

.textarea {
  border: none;
  padding: 1.5rem;
  max-width: calc(100% - 1px);
  font-size: 2rem;
  resize: vertical;
}