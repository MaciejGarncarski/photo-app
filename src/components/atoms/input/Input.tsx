import clsx from 'clsx';
import { ChangeEvent, forwardRef } from 'react';

import styles from './Input.module.scss';

type PropsTypes = {
  type?: 'text' | 'number' | 'tel' | 'email' | 'password';
  labelText: string;
  error?: string;
  optional?: boolean;
  onChange?: (ev: ChangeEvent<HTMLInputElement>) => void;
  value?: string;
};

export const Input = forwardRef<HTMLInputElement, PropsTypes>(
  (
    { type = 'text', labelText, error, optional, onChange, value, ...props },
    ref,
  ) => {
    const containerClassName = clsx(
      error && styles.containerError,
      styles.container,
    );

    const inputClassName = clsx(
      {
        [styles.inputError]: Boolean(error),
      },
      styles.input,
    );

    return (
      <div className={containerClassName}>
        <label className={styles.label} htmlFor={labelText}>
          {labelText} {optional && '(optional)'}
        </label>
        <input
          ref={ref}
          type={type}
          id={labelText}
          data-testid="input"
          className={inputClassName}
          onChange={onChange}
          value={value}
          {...props}
        />
        {error && <p className={styles.error}>{error}</p>}
      </div>
    );
  },
);
