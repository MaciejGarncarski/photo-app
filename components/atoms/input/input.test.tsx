import { screen } from '@testing-library/react';

import { render } from '@/utils/tests/utils';

import { Input } from '@/components/atoms/input/Input';

describe('Input', () => {
  describe('integration', () => {
    it('Should pass when error displayed and when has error className', () => {
      render(<Input labelText="label" optional error="max error" />);
      const errorMessage = screen.getByText(/max error/i);
      const input = screen.getByTestId(/input/i);
      expect(errorMessage).toBeInTheDocument();
      expect(input.classList).toContain('inputError');
    });
  });
});
