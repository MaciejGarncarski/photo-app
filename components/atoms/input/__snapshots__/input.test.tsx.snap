// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <label
      class="label"
      for="random"
    >
      random 
    </label>
    <input
      class="input"
      data-testid="test"
      id="random"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
